internal/controllers/clusterclass/clusterclass_controller.go

```diff
@@ -133,37 +132,74 @@ func (r *Reconciler) reconcile(ctx context.Context, clusterClass *clusterv1.Clus
 		}
 	}
 
-	// Ensure all the referenced objects are owned by the ClusterClass and that references are
-	// upgraded to the latest contract.
-	// Nb. Some external objects can be referenced multiple times in the ClusterClass. We
-	// update the API contracts of all the references but we set the owner reference on the unique
-	// external object only once.
+	// Ensure all referenced objects are owned by the ClusterClass.
+	// Nb. Some external objects can be referenced multiple times in the ClusterClass,
+	// but we only want to set the owner reference once per unique external object.
+	// For example the same KubeadmConfigTemplate could be referenced in multiple MachineDeployment
+	// classes.
 	errs := []error{}
-	patchedRefs := sets.NewString()
+	reconciledRefs := sets.NewString()
+	outdatedRefs := map[*corev1.ObjectReference]*corev1.ObjectReference{}
 	for i := range refs {
 		ref := refs[i]
 		uniqueKey := uniqueObjectRefKey(ref)
-		if err := r.reconcileExternal(ctx, clusterClass, ref, !patchedRefs.Has(uniqueKey)); err != nil {
+
+		// Continue as we only have to reconcile every referenced object once.
+		if reconciledRefs.Has(uniqueKey) {
+			continue
+		}
+
+		reconciledRefs.Insert(uniqueKey)
+
+		if err := r.reconcileExternal(ctx, clusterClass, ref); err != nil {
```

Comment:
```suggestion
		// Add the ClusterClass as owner reference to the template
		// so clusterctl move can identify all the related objects.
		if err := r.reconcileExternal(ctx, clusterClass, ref); err != nil {
```
note: I'm starting to think that we can bake in clusterctl move some specific knowledge of our types, but this is a different discussion

Reply:
Interesting idea. Should be way more efficient then setting ownerRefs everywhere (but I guess in a lot of cases we want them for delete use-cases anyway)

Reply:
Mentioned both
```go
		// Add the ClusterClass as owner reference to the templates so clusterctl move 
		// can identify all related objects and Kubernetes garbage collector deletes 
		// all referenced templates on ClusterClass deletion.
```